sound/soc/sof/pm.c

```diff
@@ -56,6 +56,10 @@ static int sof_resume(struct device *dev, bool runtime_resume)
 	if (!sof_ops(sdev)->resume || !sof_ops(sdev)->runtime_resume)
 		return 0;
 
+	/* Nothing to do if firmware never booted */
```

Comment:
that doesn't even begin to describe the issue...

Also I am not sure if this is right. don't we boot the DSP on resume?

Reply:
@plbossart yes, we do boot the DSP on resume. But if the firmware boot failed at first boot, it is likely to fail again during resume. So, isnt it better to avoid it?

Reply:
@ranj063 @plbossart Hmm, this is starting to get hard to follow and my previous patch that misused "boot_complete" is contributing to the overall difficulty. 

How about if we add an explicit enum of FW states, FW_NOT_STARTED, FW_FIRST_BOOT, FW_BOOT_OK, FW_BOOT_ERROR  and replace both first_boot and boot_complete with the enum? 

Now the above state names are maybe not the best possible, but these would already make the code more self-documenting. E.g. if state==FW_BOOT_ERROR, we can return directly here, the DSP won't respond anyways. The "if (first_boot)" is much less obvious.

Another thing is the conditions when to not retry boot because earlier errors. Maybe it's ok to skip the boot if the initial fw boot failed. But for other errors (a rare FW error during some use-case flow), I'd be more careful as a power cycle and reload FW may be a succesful recovery method in these cases. But yeah, for initial FW load (until FW_READY), for that retry is unlikely to help.

Reply:
@kv2019i agree, time for proper state machines. Thanks for suggesting this.